Support emitting function pointers in signatures

333fred opened the pull request:
@AlekseyTs @dotnet/roslyn-compiler for review.

AlekseyTs commented:
@333fred It looks like tests are failing

cston commented on src/Compilers/Core/Portable/PEWriter/Types.cs:
```diff
@@ -384,6 +384,17 @@ internal interface IPointerTypeReference : ITypeReference
     }
 
+    /// <summary>
+    /// This interface modesl the metadata representation of a pointer to a function in unmanaged memory.
```
>modesl [](start = 23, length = 6)

Typo

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -81,21 +83,50 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
+                retAndParamTypes.AsSpan()[1..]);
+        }
+
+        private static bool HasInModifier(ImmutableArray<CustomModifier> refCustomModifiers)
```
>HasInModifier [](start = 28, length = 13)

Is this distinct from `CustomModifierUtils.HasInAttributeModifier()`?

333fred replied:
Only in that I did not realize the latter existed :)

---
In reply to: [374321386](https://github.com/dotnet/roslyn/pull/41340#discussion_r374321386) [](ancestors = 374321386)

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -110,6 +141,46 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
+            else
+            {
+                _parameters = ImmutableArray<FunctionPointerParameterSymbol>.Empty;
+            }
```
Consider extracting a `static ImmutableArray<ParameterSymbol> MakeParameters(...)` method.

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -131,6 +202,9 @@ internal bool Equals(FunctionPointerMethodSymbol other, TypeCompareKind compareK
                && RefKind == other.RefKind
+               && ((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0
+                    ? RefCustomModifiers.SequenceEqual(other.RefCustomModifiers)
+                    : true)
```
Perhaps:
```
((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) != 0 ||
RefCustomModifiers.SequenceEqual(other.RefCustomModifiers))
```
Same comment for FunctionPointerParameterSymbol.

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:
```diff
@@ -8,7 +8,7 @@
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    internal sealed class FunctionPointerParameterSymbol : ParameterSymbol
+    internal sealed partial class FunctionPointerParameterSymbol : ParameterSymbol
```
>partial [](start = 20, length = 7)

Is this change needed?

333fred replied:
Not anymore, thought I was going to need to create a specific Cci adapter.

---
In reply to: [374336827](https://github.com/dotnet/roslyn/pull/41340#discussion_r374336827) [](ancestors = 374336827)

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -5,14 +5,16 @@
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    internal sealed class FunctionPointerMethodSymbol : MethodSymbol
+    internal sealed partial class FunctionPointerMethodSymbol : MethodSymbol
```
>partial [](start = 20, length = 7)

Is this needed?

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/Types.cs:
```diff
@@ -384,6 +384,17 @@ internal interface IPointerTypeReference : ITypeReference
+        /// <summary>
+        /// The signature of the function located at the target memory address.
+        /// </summary>
+        IMethodReference Signature { get; }
```
>IMethodReference [](start = 8, length = 16)

I think this should be ```ISignature```

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed partial class FunctionPointerTypeSymbol : Cci.IFunctionPointerTypeReference
```
>FunctionPointerTypeSymbol [](start = 34, length = 25)

It looks like there is a typo in the name of the file - "FunctionPointerTypeSymbol**s**Adapter"

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -12,19 +12,21 @@
+    // PROTOTYPE(func-ptr): Match msvc's emitting of custom modifiers for calling convention
+    internal sealed partial class FunctionPointerTypeSymbol : TypeSymbol
     {
         public static FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbolFromSource(FunctionPointerTypeSyntax syntax, Binder typeBinder, DiagnosticBag diagnostics, ConsList<TypeSymbol> basesBeingResolved, bool suppressUseSiteDiagnostics)
```
>FunctionPointerTypeSymbolFromSource [](start = 54, length = 35)

Consider shortening name to `Create` or `CreateFromSource`. Similar comment for `Create...FromMetadata` below.

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/MetadataVisitor.cs:
```diff
@@ -465,6 +465,17 @@ public virtual void Visit(IPointerTypeReference pointerTypeReference)
 
+        public virtual void Visit(IFunctionPointerTypeReference functionPointerTypeReference)
+        {
+            this.Visit(functionPointerTypeReference.Signature.RefCustomModifiers);
```
>RefCustomModifiers [](start = 62, length = 18)

What about ReturnValueCustomModifiers?

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -758,5 +647,48 @@ class C
+                    else
+                    {
+                        Assert.True(param1.Type.Equals(param2.Type, TypeCompareKind.ConsiderEverything));
+                    }
```
```
Assert.NotEqual(
    ReferenceEquals(param1, param2),
    param1.Type.Equals(param2.Type, TypeCompareKind.ConsiderEverything));
```

333fred replied:
I think the way it's written currently is easier to understand and diagnose if something went wrong. I'm going to leave it as is.

---
In reply to: [374345105](https://github.com/dotnet/roslyn/pull/41340#discussion_r374345105) [](ancestors = 374345105)

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -147,6 +147,26 @@ public override void VisitPointerType(IPointerTypeSymbol symbol)
             AddPunctuation(SyntaxKind.AsteriskToken);
         }
 
+        public override void VisitFunctionPointerType(IFunctionPointerTypeSymbol symbol)
```
>VisitFunctionPointerType [](start = 29, length = 24)

Are we testing this?

333fred replied:
No, I just added it for ease of debugging (GetDebuggerDisplay uses it). I'll add a prototype, as the formatting is kinda wonky and I'm not planning on addressing this until I address the public API in general.

---
In reply to: [374345748](https://github.com/dotnet/roslyn/pull/41340#discussion_r374345748) [](ancestors = 374345748)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs:
```diff
@@ -4,6 +4,7 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using Microsoft.Cci;
```
I think this namespace was not imported intentionally.

333fred replied:
It is, `CallingConvention` is a Cci type. If you prefer I can qualify that one instance and remove the import.

---
In reply to: [374347237](https://github.com/dotnet/roslyn/pull/41340#discussion_r374347237) [](ancestors = 374347237)

AlekseyTs replied:
> If you prefer I can qualify that one instance and remove the import.

Yes, I would prefer we do this as we are doing for other Cci type references in this file.

---
In reply to: [374393583](https://github.com/dotnet/roslyn/pull/41340#discussion_r374393583) [](ancestors = 374393583,374347237)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
+                case SignatureTypeCode.FunctionPointer:
+                    var signatureHeader = ppSig.ReadSignatureHeader();
+                    var parameters = DecodeSignatureParametersOrThrow(ref ppSig, signatureHeader, typeParameterCount: out int typeParamCount, shouldProcessAllBytes: false);
+                    Debug.Assert(typeParamCount == 0);
```
>Debug.Assert(typeParamCount == 0); [](start = 20, length = 34)

I think we should do a check and ```throw new UnsupportedSignatureContent()``` otherwise.

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
+                    {
+                        SignatureCallingConvention.Default => Cci.CallingConvention.Default,
+                        SignatureCallingConvention.CDecl => Cci.CallingConvention.CDecl,
+                        SignatureCallingConvention.FastCall => Cci.CallingConvention.FastCall,
```
>Cci.CallingConvention.FastCall [](start = 63, length = 30)

It looks like we will not be able to serialize a type like that. Is there a reason for that? In general, we should be able to gracefully handle everything we import, but it looks like MetadatWriter is going to crash on this convention.

333fred replied:
As currently spec'd, C# function pointers don't have fastcall. I'll raise the question and make this throw for now.

---
In reply to: [374388482](https://github.com/dotnet/roslyn/pull/41340#discussion_r374388482) [](ancestors = 374388482)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
+                        SignatureCallingConvention.FastCall => Cci.CallingConvention.FastCall,
+                        SignatureCallingConvention.StdCall => Cci.CallingConvention.Standard,
+                        SignatureCallingConvention.ThisCall => Cci.CallingConvention.ThisCall,
+                        SignatureCallingConvention.VarArgs => Cci.CallingConvention.ExtraArguments,
```
>Cci.CallingConvention.ExtraArguments [](start = 62, length = 36)

Same comment

333fred replied:
This one should probably just throw, I don't think we're actually planning on supporting varags function pointers in C#.

---
In reply to: [374388711](https://github.com/dotnet/roslyn/pull/41340#discussion_r374388711) [](ancestors = 374388711)

AlekseyTs replied:
> This one should probably just throw, I don't think we're actually planning on supporting varags function pointers in C#.

What we should be able to import/emit and what you can define in source code is not necessary the same thing. I think that up until the language is getting involved and symbol model is able to represent the metadata, we shouldn't fail

---
In reply to: [374392051](https://github.com/dotnet/roslyn/pull/41340#discussion_r374392051) [](ancestors = 374392051,374388711)

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs:
```diff
@@ -0,0 +1,29 @@
+    internal sealed partial class FunctionPointerTypeSymbol : Cci.IFunctionPointerTypeReference
+    {
+        IMethodReference IFunctionPointerTypeReference.Signature => Signature;
+        void IReference.Dispatch(Cci.MetadataVisitor visitor) => visitor.Visit((IFunctionPointerTypeReference)this);
```
>IReference.Dispatch [](start = 13, length = 19)

It feels like we should also implement two other APIs of IReference

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:
```diff
@@ -162,9 +162,11 @@ private TypeSymbol DecodeType(TypeSymbol type)
                 case SymbolKind.DynamicType:
                 case SymbolKind.TypeParameter:
                 case SymbolKind.PointerType:
+                case SymbolKind.FunctionPointer:
```
>case SymbolKind.FunctionPointer: [](start = 16, length = 32)

This doesn't feel right.

AlekseyTs commented:
>     internal struct TupleTypeDecoder

It feels like other decoders should also be adjusted: nullable, dynamic, etc.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:63 in c83dd64. [](commit_id = c83dd642d5f722a90583a2058438a16a8fc562ea, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,56 @@
+    internal sealed class FunctionPointerTypeSymbol : TypeSymbol, IFunctionPointerTypeSymbol
+    {
+        private readonly Symbols.FunctionPointerTypeSymbol _underlying;
+        private IMethodSymbol? _lazySignatureSymbol;
```
>private IMethodSymbol? _lazySignatureSymbol; [](start = 8, length = 44)

I am not sure if this cache really needed. I think each symbol already caches its public representation

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs:
```diff
@@ -0,0 +1,56 @@
+        protected override ITypeSymbol WithNullableAnnotation(CodeAnalysis.NullableAnnotation nullableAnnotation)
+        {
+            Debug.Assert(nullableAnnotation != this.NullableAnnotation);
+            return new FunctionPointerTypeSymbol(_underlying, nullableAnnotation);
```
>return [](start = 12, length = 6)

```Debug.Assert(nullableAnnotation != _underlying.DefaultNullableAnnotation);```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -83,24 +84,22 @@ internal override bool Equals(TypeSymbol t2, TypeCompareKind compareKind, IReadO
         {
-            // PROTOTYPE(func-ptr): Implement
-            throw new NotImplementedException();
+            return CreateITypeSymbol(CodeAnalysis.NullableAnnotation.None);
```
>CodeAnalysis.NullableAnnotation.None [](start = 37, length = 36)

DefaultNullableAnnotation?

AlekseyTs replied:
>return CreateITypeSymbol [](start = 12, length = 24)

I think the code reuse doesn't give us a lot of benefits, but disallows us to assert that we properly use internal cash for public symbols.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -83,24 +84,22 @@ internal override bool Equals(TypeSymbol t2, TypeCompareKind compareKind, IReadO
         {
-            // PROTOTYPE(func-ptr): Implement
-            throw new NotImplementedException();
+            return new PublicModel.FunctionPointerTypeSymbol(this, nullableAnnotation);
```
>return [](start = 12, length = 6)

```Debug.Assert(nullableAnnotation != DefaultNullableAnnotation);```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -81,21 +83,50 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
+        {
+            Debug.Assert(retAndParamTypes.Length > 0);
+
+            ParamInfo<TypeSymbol> retInfo = retAndParamTypes[0];
```
>ParamInfo<TypeSymbol> retInfo = retAndParamTypes[0] [](start = 12, length = 51)

Why can't we do all this return type related work within the constructor?

333fred replied:
We could. But given that Chuck just asked me to move what _was_ in the constructor into a helper method, I'm going to leave this where it is.

---
In reply to: [374405675](https://github.com/dotnet/roslyn/pull/41340#discussion_r374405675) [](ancestors = 374405675)

AlekseyTs replied:
> We could. But given that Chuck just asked me to move what was in the constructor into a helper method, I'm going to leave this where it is.

I am not sure how one is related to the other. I think, from the API design perspective, it would be better if constructor were taking all data in raw form, or all data in "preprocessed" form.


---
In reply to: [374952548](https://github.com/dotnet/roslyn/pull/41340#discussion_r374952548) [](ancestors = 374952548,374405675)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -0,0 +1,342 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
+{
+    public class CodeGenFunctionPointersTests : CSharpTestBase
```
>CodeGenFunctionPointersTests [](start = 17, length = 28)

It might be useful to have targeted tests for metadata import from IL to ensure that things end up in the right places.

AlekseyTs commented:
Done with review pass (iteration 2)

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -3695,6 +3701,24 @@ private void SerializeTypeReference(SignatureTypeEncoder encoder, ITypeReference
+                    SerializeReturnValueAndParameters(signatureEncoder, signature, varargParameters: ImmutableArray<IParameterTypeInformation>.Empty);
+                    return;
+
+                    static SignatureCallingConvention convertCallingConvention(CallingConvention originalConvention)
```
>SignatureCallingConvention [](start = 27, length = 26)

It feels like translating between two enums is just another hoop to jump over. Would it be possible to remove the Cci version of the enum and switch to using ```SignatureCallingConvention```, which, I assume, is a public type and could be used in our public APIs as well?

333fred commented:
>     internal struct TupleTypeDecoder

I'll put a prototype. I was specifically avoiding them in this pr.

---
In reply to: [581664963](https://github.com/dotnet/roslyn/pull/41340#issuecomment-581664963) [](ancestors = 581664963)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:63 in c83dd64. [](commit_id = c83dd642d5f722a90583a2058438a16a8fc562ea, deletion_comment = False)

333fred commented:
@AlekseyTs @cston addressed feedback.

---
In reply to: [581682378](https://github.com/dotnet/roslyn/pull/41340#issuecomment-581682378) [](ancestors = 581682378)

cston commented on src/Compilers/Core/Portable/PEWriter/Members.cs:
```diff
@@ -70,6 +70,23 @@ internal enum CallingConvention
+            | SignatureCallingConvention.VarArgs;
+
+        internal static CallingConvention FromSignatureConvention(this SignatureCallingConvention convention)
+            => (CallingConvention)(convention & SignatureMask);
```
>convention & SignatureMask [](start = 35, length = 26)

Are we silently ignoring bits outside `SignatureMask`? Should we `throw UnsupportedSignatureContent()` in those cases instead?

cston commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -319,6 +319,18 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
+                        throw new UnsupportedSignatureContent();
+                    }
+
+                    typeSymbol = MakeFunctionPointerTypeSymbol(Cci.CallingConventionUtils.FromSignatureConvention(signatureHeader.CallingConvention), ImmutableArray.Create(parameters));
```
>Cci.CallingConventionUtils.FromSignatureConvention(signatureHeader.CallingConvention) [](start = 63, length = 85)

`signatureHeader.CallingConvention.FromSignatureConvention()`

333fred replied:
Actually, there aren't that many here, I'll just do it.

---
In reply to: [375578610](https://github.com/dotnet/roslyn/pull/41340#discussion_r375578610) [](ancestors = 375578610,375063146)

333fred replied:
> there aren't that many here

But there _are_ conflicting names between Cci. and SRM., so I'll leave it.

---
In reply to: [375582556](https://github.com/dotnet/roslyn/pull/41340#discussion_r375582556) [](ancestors = 375582556,375578610,375063146)

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -81,21 +83,77 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
+                        false => RefKind.None,
+                        true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => RefKind.In,
+                        true => RefKind.Ref,
+                    };
```
Consider extracting a helper method for use here and in `CreateFromMetadata()` above.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -81,21 +83,77 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
+                retAndParamTypes.AsSpan()[1..]);
+        }
+
+        private static ImmutableArray<FunctionPointerParameterSymbol> MakeParametersFromMetadata(ReadOnlySpan<ParamInfo<TypeSymbol>> parameterTypes, FunctionPointerMethodSymbol parent)
```
>MakeParametersFromMetadata [](start = 70, length = 26)

It looks like there is only one call site, consider making this a local function
